subsys/bluetooth/host/classic/hfp_ag.c

```diff
@@ -385,256 +357,1118 @@ static int hfp_ag_send_data(struct bt_hfp_ag *ag, bt_hfp_ag_tx_cb_t cb, void *us
+
+	if (state == BT_HFP_DISCONNECTING) {
+		int err = bt_rfcomm_dlc_disconnect(&ag->rfcomm_dlc);
+
+		if (err) {
+			LOG_ERR("Fail to disconnect DLC %p", &ag->rfcomm_dlc);
+		}
+	}
+}
+
+static void hfp_ag_close_sco(struct bt_hfp_ag *ag)
+{
+	struct bt_conn *sco;
+	int call_count;
+
+	LOG_DBG("");
+
+	hfp_ag_lock(ag);
+	call_count = get_none_released_calls(ag);
+	if (call_count > 0) {
+		LOG_INF("Do not close sco because not all calls are released");
+		hfp_ag_unlock(ag);
+		return;
+	}
+
+	sco = ag->sco_chan.sco;
+	ag->sco_chan.sco = NULL;
+	hfp_ag_unlock(ag);
+	if (sco != NULL) {
+		LOG_DBG("Disconnect sco %p", sco);
+		bt_conn_disconnect(sco, BT_HCI_ERR_LOCALHOST_TERM_CONN);
+	}
+}
+
+static void ag_reject_call(struct bt_hfp_ag_call *call)
+{
+	struct bt_hfp_ag *ag;
+
+	ag = call->ag;
+
+	if (bt_ag && bt_ag->reject) {
+		bt_ag->reject(call);
+	}
+
+	bt_hfp_ag_set_call_state(call, BT_HFP_CALL_TERMINATE);
+
+	if (!atomic_test_bit(ag->flags, BT_HFP_AG_AUDIO_CONN)) {
+		LOG_DBG("It is not audio connection");
+		hfp_ag_close_sco(ag);
+	}
+}
+
+static void ag_terminate_call(struct bt_hfp_ag_call *call)
+{
+	struct bt_hfp_ag *ag;
+
+	ag = call->ag;
+
+	if (bt_ag && bt_ag->terminate) {
+		bt_ag->terminate(call);
+	}
+
+	bt_hfp_ag_set_call_state(call, BT_HFP_CALL_TERMINATE);
+
+	if (!atomic_test_bit(ag->flags, BT_HFP_AG_AUDIO_CONN)) {
+		LOG_DBG("It is not audio connection");
+		hfp_ag_close_sco(ag);
+	}
+}
+
+static int hfp_ag_send(struct bt_hfp_ag *ag, struct bt_ag_tx *tx)
+{
+	int err = bt_rfcomm_dlc_send(&ag->rfcomm_dlc, tx->buf);
+
+	if (err < 0) {
+		net_buf_unref(tx->buf);
+	}
+
+	return err;
+}
+
+static void bt_ag_tx_work(struct k_work *work)
+{
+	struct k_work_delayable *dwork = k_work_delayable_from_work(work);
+	struct bt_hfp_ag *ag = CONTAINER_OF(dwork, struct bt_hfp_ag, tx_work);
+	sys_snode_t *node;
+	struct bt_ag_tx *tx;
+	bt_hfp_state_t state;
+
+	hfp_ag_lock(ag);
+	state = ag->state;
+	if ((state == BT_HFP_DISCONNECTED) || (state == BT_HFP_DISCONNECTING)) {
+		LOG_ERR("AG %p is not connected", ag);
+		goto unlock;
+	}
+
+	node = sys_slist_peek_head(&ag->tx_pending);
+	if (!node) {
+		LOG_DBG("No pending tx");
+		goto unlock;
+	}
+
+	tx = CONTAINER_OF(node, struct bt_ag_tx, node);
+
+	if (!atomic_test_and_set_bit(ag->flags, BT_HFP_AG_TX_ONGOING)) {
+		LOG_DBG("AG %p sending tx %p", ag, tx);
+		int err = hfp_ag_send(ag, tx);
+
+		if (err < 0) {
+			LOG_ERR("Rfcomm send error :(%d)", err);
+			sys_slist_find_and_remove(&ag->tx_pending, &tx->node);
+			tx->err = err;
+			k_fifo_put(&ag_tx_notify, tx);
+			/* Clear the tx ongoing flag */
+			if (!atomic_test_and_clear_bit(ag->flags, BT_HFP_AG_TX_ONGOING)) {
+				LOG_WRN("tx ongoing flag is not set");
+			}
+			/* Due to the work is failed, restart the tx work */
+			k_work_reschedule(&ag->tx_work, K_NO_WAIT);
+		}
+	}
+
+unlock:
+	hfp_ag_unlock(ag);
+}
+
+static void skip_space(struct net_buf *buf)
+{
+	while ((buf->len > 0) && (buf->data[0] == ' ')) {
+		(void)net_buf_pull(buf, 1);
+	}
```

Comment:
This would be more efficiently implemented if you had a counter and only did a single `net_buf_pull(buf, count)` call. This is because `net_buf_pull` updates several different variables, whereas incrementing a counter is just one variable.

Reply:
Updated.